File: src/com/maddyhome/idea/vim/regexp/Magic.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2013 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.regexp;

public class Magic {
  public static final int AMP = '&' - 256;
  public static final int AT = '@' - 256;
  public static final int DOLLAR = '$' - 256;
  public static final int DOT = '.' - 256;
  public static final int EQUAL = '=' - 256;
  public static final int GREATER = '>' - 256;
  public static final int HAT = '^' - 256;
  public static final int LBRACE = '[' - 256;
  public static final int LCURLY = '{' - 256;
  public static final int LESS = '<' - 256;
  public static final int LPAREN = '(' - 256;
  public static final int PERCENT = '%' - 256;
  public static final int PIPE = '|' - 256;
  public static final int PLUS = '+' - 256;
  public static final int QUESTION = '?' - 256;
  public static final int RPAREN = ')' - 256;
  public static final int STAR = '*' - 256;
  public static final int TILDE = '~' - 256;
  public static final int UNDER = '_' - 256;
  public static final int N0 = '0' - 256;
  public static final int N1 = '1' - 256;
  public static final int N2 = '2' - 256;
  public static final int N3 = '3' - 256;
  public static final int N4 = '4' - 256;
  public static final int N5 = '5' - 256;
  public static final int N6 = '6' - 256;
  public static final int N7 = '7' - 256;
  public static final int N8 = '8' - 256;
  public static final int N9 = '9' - 256;
  public static final int a = 'a' - 256;
  public static final int A = 'A' - 256;
  public static final int c = 'c' - 256;
  public static final int C = 'C' - 256;
  public static final int d = 'd' - 256;
  public static final int D = 'D' - 256;
  public static final int f = 'f' - 256;
  public static final int F = 'F' - 256;
  public static final int h = 'h' - 256;
  public static final int H = 'H' - 256;
  public static final int i = 'i' - 256;
  public static final int I = 'I' - 256;
  public static final int k = 'k' - 256;
  public static final int K = 'K' - 256;
  public static final int l = 'l' - 256;
  public static final int L = 'L' - 256;
  public static final int m = 'm' - 256;
  public static final int M = 'M' - 256;
  public static final int n = 'n' - 256;
  public static final int N = 'N' - 256;
  public static final int o = 'o' - 256;
  public static final int O = 'O' - 256;
  public static final int p = 'p' - 256;
  public static final int P = 'P' - 256;
  public static final int s = 's' - 256;
  public static final int S = 'S' - 256;
  public static final int u = 'u' - 256;
  public static final int U = 'U' - 256;
  public static final int v = 'v' - 256;
  public static final int V = 'V' - 256;
  public static final int w = 'w' - 256;
  public static final int W = 'W' - 256;
  public static final int x = 'x' - 256;
  public static final int X = 'X' - 256;
  public static final int z = 'z' - 256;

  /*
  * Magic characters have a special meaning, they don't match literally.
  * Magic characters are negative.  This separates them from literal characters
  * (possibly multi-byte).  Only ASCII characters can be Magic.
  */
  public static int Magic(int x) {
    return (x - 256);
  }

  public static int un_Magic(int x) {
    return (x + 256);
  }

  public static boolean is_Magic(int x) {
    return (x < 0);
  }

  public static int no_Magic(int x) {
    if (is_Magic(x)) {
      return un_Magic(x);
    }
    return x;
  }

  public static int toggle_Magic(int x) {
    if (is_Magic(x)) {
      return un_Magic(x);
    }
    return Magic(x);
  }
}

File: vim-engine/src/main/kotlin/com/maddyhome/idea/vim/regexp/Magic.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.regexp

@Deprecated("Remove once old regex engine is removed")
object Magic {
  const val AMP: Int = '&'.code - 256
  const val AT: Int = '@'.code - 256
  const val DOLLAR: Int = '$'.code - 256
  const val DOT: Int = '.'.code - 256
  const val EQUAL: Int = '='.code - 256
  const val GREATER: Int = '>'.code - 256
  const val HAT: Int = '^'.code - 256
  const val LBRACE: Int = '['.code - 256
  const val LCURLY: Int = '{'.code - 256
  const val LESS: Int = '<'.code - 256
  const val LPAREN: Int = '('.code - 256
  const val PERCENT: Int = '%'.code - 256
  const val PIPE: Int = '|'.code - 256
  const val PLUS: Int = '+'.code - 256
  const val QUESTION: Int = '?'.code - 256
  const val RPAREN: Int = ')'.code - 256
  const val STAR: Int = '*'.code - 256
  const val TILDE: Int = '~'.code - 256
  const val UNDER: Int = '_'.code - 256
  const val N0: Int = '0'.code - 256
  const val N1: Int = '1'.code - 256
  const val N2: Int = '2'.code - 256
  const val N3: Int = '3'.code - 256
  const val N4: Int = '4'.code - 256
  const val N5: Int = '5'.code - 256
  const val N6: Int = '6'.code - 256
  const val N7: Int = '7'.code - 256
  const val N8: Int = '8'.code - 256
  const val N9: Int = '9'.code - 256
  const val a: Int = 'a'.code - 256
  const val A: Int = 'A'.code - 256
  const val c: Int = 'c'.code - 256
  const val C: Int = 'C'.code - 256
  const val d: Int = 'd'.code - 256
  const val D: Int = 'D'.code - 256
  const val f: Int = 'f'.code - 256
  const val F: Int = 'F'.code - 256
  const val h: Int = 'h'.code - 256
  const val H: Int = 'H'.code - 256
  const val i: Int = 'i'.code - 256
  const val I: Int = 'I'.code - 256
  const val k: Int = 'k'.code - 256
  const val K: Int = 'K'.code - 256
  const val l: Int = 'l'.code - 256
  const val L: Int = 'L'.code - 256
  const val m: Int = 'm'.code - 256
  const val M: Int = 'M'.code - 256
  const val n: Int = 'n'.code - 256
  const val N: Int = 'N'.code - 256
  const val o: Int = 'o'.code - 256
  const val O: Int = 'O'.code - 256
  const val p: Int = 'p'.code - 256
  const val P: Int = 'P'.code - 256
  const val s: Int = 's'.code - 256
  const val S: Int = 'S'.code - 256
  const val u: Int = 'u'.code - 256
  const val U: Int = 'U'.code - 256
  const val v: Int = 'v'.code - 256
  const val V: Int = 'V'.code - 256
  const val w: Int = 'w'.code - 256
  const val W: Int = 'W'.code - 256
  const val x: Int = 'x'.code - 256
  const val X: Int = 'X'.code - 256
  const val z: Int = 'z'.code - 256

  /*
 * Magic characters have a special meaning, they don't match literally.
 * Magic characters are negative.  This separates them from literal characters
 * (possibly multi-byte).  Only ASCII characters can be Magic.
 */
  fun magic(x: Int): Int {
    return x - 256
  }

  fun un_Magic(x: Int): Int {
    return x + 256
  }

  fun is_Magic(x: Int): Boolean {
    return x < 0
  }

  fun no_Magic(x: Int): Int {
    return if (is_Magic(x)) {
      un_Magic(x)
    } else {
      x
    }
  }

  fun toggle_Magic(x: Int): Int {
    return if (is_Magic(x)) {
      un_Magic(x)
    } else {
      magic(x)
    }
  }
}
